tests/bgp/bgp_helpers.py

```diff
@@ -514,3 +522,210 @@ def get_default_action():
+    mg_facts = duthost.get_extended_minigraph_facts(tbinfo)
+    return mg_facts['minigraph_ptf_indices'][port]
+
+
+def get_eth_port(duthost, tbinfo):
+    """
+    Get ethernet port that connects to T0 VM
+    """
+    mg_facts = duthost.get_extended_minigraph_facts(tbinfo)
+    t0_vm = [vm_name for vm_name in mg_facts['minigraph_devices'].keys() if vm_name.endswith('T0')][0]
+    port = duthost.shell("show ip interface | grep -w {} | awk '{{print $1}}'".format(t0_vm))['stdout']
+    return port
+
+
+def get_vm_offset(duthost, nbrhosts, tbinfo):
+    """
+    Get port offset of exabgp and ptf receive port
+    """
+    logging.info("get_vm_offset ---------")
+    vm_name = random.choice([vm_name for vm_name in nbrhosts.keys() if vm_name.endswith('T0')])
+    ptf_recv_port = get_ptf_recv_port(duthost, vm_name, tbinfo)
+    port_offset = tbinfo['topo']['properties']['topology']['VMs'][vm_name]['vm_offset']
+    logging.info("vm_offset of {} is: {}".format(vm_name, port_offset))
+    return port_offset, ptf_recv_port
+
+
+def get_exabgp_port(duthost, nbrhosts, tbinfo, exabgp_base_port):
+    """
+    Get exabgp port and ptf receive port
+    """
+    port_offset, ptf_recv_port = get_vm_offset(duthost, nbrhosts, tbinfo)
+    return exabgp_base_port + port_offset, ptf_recv_port
+
+
+def get_vm_name(tbinfo, vm_level='T2'):
+    """
+    Get vm name, default return value would be T2 VM name
+    """
+    for vm in tbinfo['topo']['properties']['topology']['VMs'].keys():
+        if vm[-2:] == vm_level:
+            return vm
+
+
+def get_t2_ptf_intfs(mg_facts):
+    """
+    Get ptf interface list that connect with T2 VMs
+    """
+    t2_ethernets = []
+    for k, v in mg_facts["minigraph_neighbors"].items():
+        if v['name'][-2:] == 'T2':
+            t2_ethernets.append(k)
+
+    ptf_interfaces = []
+    for port in t2_ethernets:
+        ptf_interfaces.append(mg_facts['minigraph_ptf_indices'][port])
+    return ptf_interfaces
+
+
+def get_bgp_neighbor_ip(duthost, vm_name, vrf=DEFAULT):
+    """
+    Get ipv4 and ipv6 bgp neighbor ip addresses
+    """
+    if vrf == DEFAULT:
+        cmd_v4 = "show ip interface | grep -w {} | awk '{{print $5}}'"
+        cmd_v6 = "show ipv6 interface | grep -w {} | awk '{{print $5}}'"
+        bgp_neighbor_ip = duthost.shell(cmd_v4.format(vm_name))['stdout']
+        bgp_neighbor_ipv6 = duthost.shell(cmd_v6.format(vm_name))['stdout']
+    else:
+        cmd_v4 = "show ip interface | grep -w {} | awk '{{print $7}}' | sed 's/)//g'"
+        cmd_v6 = "show ipv6 interface | grep -w {} | awk '{{print $7}}' | sed 's/)//g'"
+        bgp_neighbor_ip = duthost.shell(cmd_v4.format(vm_name))['stdout'][1:-1]
+        bgp_neighbor_ipv6 = duthost.shell(cmd_v6.format(vm_name))['stdout'][1:-1]
+    logging.info("BGP neighbor of {} is {}".format(vm_name, bgp_neighbor_ip))
+    logging.info("IPv6 BGP neighbor of {} is {}".format(vm_name, bgp_neighbor_ipv6))
+
```

Comment:
There is already a function to get bgp neighbor ips, something like:
bgp_facts = duthost.bgp_facts()['ansible_facts']
bgp_facts['bgp_neighbors']

Would this method can be use in your case ?

Reply:
Hi @jcaiMR ,
Checked with this method, it could not list the bgp neighbors under user defined vrf:
(Pdb) duthost.bgp_facts()
Thursday 25 May 2023  12:56:05 +0300 (0:01:45.335)       0:24:46.708 ********** 
{u'invocation': {u'module_args': {u'instance_id': None, u'num_npus': 1}}, 'failed': False, 'changed': False, '_ansible_no_log': False, u'ansible_facts': {u'bgp_statistics': {u'ipv4_idle': 0, u'ipv6_idle': 0, u'ipv4_admin_down': 0, u'ipv6_admin_down': 0, u'ipv6': 0, u'ipv4': 0}, **u'bgp_neighbors**': {}}}

c-panther-01# show ip bgp vrf Vrf1 summary

IPv4 Unicast Summary (VRF Vrf1):
BGP router identifier 10.0.0.62, local AS number 65100 vrf-id 417
BGP table version 11346
RIB entries 12853, using 2310 KiB of memory
Peers 48, using 34 MiB of memory
Peer groups 2, using 128 bytes of memory

Neighbor        V         AS   MsgRcvd   MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd   PfxSnt Desc
10.0.0.1        4      65200      3262        78        0    0    0 00:03:38         6370        4 ARISTA01T2
10.0.0.5        4      65200      3262        79        0    0    0 00:03:38         6370        4 ARISTA03T2
10.0.0.9        4      65200      3262        79        0    0    0 00:03:39         6370        4 ARISTA05T2
10.0.0.13       4      65200      3262        78        0    0    0 00:03:37         6370        4 ARISTA07T2
10.0.0.17       4      65200      3262        78        0    0    0 00:03:38         6370        4 ARISTA09T2
10.0.0.21       4      65200      3262        81        0    0    0 00:03:37         6370        4 ARISTA11T2
10.0.0.25       4      65200      3262        78        0    0    0 00:03:38         6370        4 ARISTA13T2
10.0.0.29       4      65200      3262        78        0    0    0 00:03:38         6370        4 ARISTA15T2
10.0.0.33       4      64001        80        79        0    0    0 00:03:39            4        4 ARISTA01T0
10.0.0.35       4      64002        78        78        0    0    0 00:03:38            3        4 ARISTA02T0
10.0.0.37       4      64003        79        78        0    0    0 00:03:38            4        4 ARISTA03T0
10.0.0.39       4      64004        78        78        0    0    0 00:03:38            3        4 ARISTA04T0
10.0.0.41       4      64005        78        78        0    0    0 00:03:38            3        4 ARISTA05T0
10.0.0.43       4      64006        78        79        0    0    0 00:03:39            3        4 ARISTA06T0
10.0.0.45       4      64007        78        78        0    0    0 00:03:37            3        4 ARISTA07T0
10.0.0.47       4      64008        78        78        0    0    0 00:03:38            3        4 ARISTA08T0
10.0.0.49       4      64009        78        78        0    0    0 00:03:38            3        4 ARISTA09T0
10.0.0.51       4      64010        78        78        0    0    0 00:03:37            3        4 ARISTA10T0
10.0.0.53       4      64011        78        78        0    0    0 00:03:38            3        4 ARISTA11T0
10.0.0.55       4      64012        80        78        0    0    0 00:03:38            5        4 ARISTA12T0
10.0.0.57       4      64013        78        78        0    0    0 00:03:38            3        4 ARISTA13T0
10.0.0.59       4      64014        78        78        0    0    0 00:03:38            3        4 ARISTA14T0
10.0.0.61       4      64015        78        78        0    0    0 00:03:38            3        4 ARISTA15T0
10.0.0.63       4      64016        78        78        0    0    0 00:03:38            3        4 ARISTA16T0
fc00::2         4      65200      3260       129        0    0    0 00:03:30        NoNeg    NoNeg ARISTA01T2
fc00::a         4      65200      3263       190        0    0    0 00:03:39        NoNeg    NoNeg ARISTA03T2
fc00::12        4      65200      3260       129        0    0    0 00:03:30        NoNeg    NoNeg ARISTA05T2
fc00::1a        4      65200      3260       129        0    0    0 00:03:31        NoNeg    NoNeg ARISTA07T2
fc00::22        4      65200      3260       129        0    0    0 00:03:32        NoNeg    NoNeg ARISTA09T2
fc00::2a        4      65200      3260       130        0    0    0 00:03:32        NoNeg    NoNeg ARISTA11T2
fc00::32        4      65200      3260       129        0    0    0 00:03:31        NoNeg    NoNeg ARISTA13T2
fc00::3a        4      65200      3262       189        0    0    0 00:03:38        NoNeg    NoNeg ARISTA15T2
fc00::42        4      64001        75       123        0    0    0 00:03:28        NoNeg    NoNeg ARISTA01T0
fc00::46        4      64002        76       129        0    0    0 00:03:30        NoNeg    NoNeg ARISTA02T0
fc00::4a        4      64003        75       128        0    0    0 00:03:29        NoNeg    NoNeg ARISTA03T0
fc00::4e        4      64004        75       128        0    0    0 00:03:29        NoNeg    NoNeg ARISTA04T0
fc00::52        4      64005        75       128        0    0    0 00:03:29        NoNeg    NoNeg ARISTA05T0
fc00::56        4      64006        75       123        0    0    0 00:03:28        NoNeg    NoNeg ARISTA06T0
fc00::5a        4      64007        75       123        0    0    0 00:03:29        NoNeg    NoNeg ARISTA07T0
fc00::5e        4      64008        75       128        0    0    0 00:03:29        NoNeg    NoNeg ARISTA08T0
fc00::62        4      64009        76       129        0    0    0 00:03:30        NoNeg    NoNeg ARISTA09T0
fc00::66        4      64010        75       123        0    0    0 00:03:29        NoNeg    NoNeg ARISTA10T0
fc00::6a        4      64011        76       129        0    0    0 00:03:30        NoNeg    NoNeg ARISTA11T0
fc00::6e        4      64012        75       128        0    0    0 00:03:29        NoNeg    NoNeg ARISTA12T0
fc00::72        4      64013        75       128        0    0    0 00:03:29        NoNeg    NoNeg ARISTA13T0
fc00::76        4      64014        76       129        0    0    0 00:03:30        NoNeg    NoNeg ARISTA14T0
fc00::7a        4      64015        76       129        0    0    0 00:03:30        NoNeg    NoNeg ARISTA15T0
fc00::7e        4      64016        76       129        0    0    0 00:03:30        NoNeg    NoNeg ARISTA16T0

Total number of neighbors 48

Reply:
I see, thanks for the information.